File: ds4-gateway/recovery_omlx_enrollment_test.py
import copy
import importlib.util
import json
from pathlib import Path
import tempfile
import unittest
from unittest.mock import patch

spec=importlib.util.spec_from_file_location('omlx_enrollment',Path(__file__).with_name('recovery-omlx-enrollment.py'))
m=importlib.util.module_from_spec(spec);spec.loader.exec_module(m)


class OmlxEnrollmentTests(unittest.TestCase):
    def setUp(self):
        self.temp=tempfile.TemporaryDirectory();self.addCleanup(self.temp.cleanup)
        self.root=Path(self.temp.name);self.destination=self.root/'enrollment'
        self.expected={'worker_id':'local','route':{'id':'local','url':'http://127.0.0.1:8013/v1'},
                       'target':{'kind':'omlx-local','root':str(self.root),'url':'http://127.0.0.1:8013/v1','api_key_file':str(self.root/'key')},
                       'launcher':str(self.root/'start.sh'),'profile_files':[],
                       'model':'fixture','context_length':400000,'concurrency':1}
        self.config={'root':str(self.root),'binary':'/fixture/python','port':8013,'command_sha256':'a'*64,
                     'api_key_file':str(self.root/'key'),'launcher':self.expected['launcher'],'profile_files':[],'start_stopped':False}
        self.inspection={'machine':'a'*64,'profile':'b'*64,'instance':'c'*32,'active':True,'listener':True,'stopped':False,'fault':None}
        self.metadata={'model':'fixture','context_length':400000,'configured_concurrency':1}

    def run_capture(self,**kwargs):
        return m.materialize(self.destination,self.expected,capture=lambda _:copy.deepcopy(self.config),
                             inspect=kwargs.get('inspect',lambda _:copy.deepcopy(self.inspection)),
                             read_metadata=kwargs.get('metadata',lambda _:copy.deepcopy(self.metadata)))

    def test_capture_pins_stable_native_identity_and_is_idempotent_without_actions(self):
        with patch.object(m.omlx.os,'kill') as kill,patch.object(m.omlx,'start') as start:
            first=self.run_capture();second=self.run_capture()
            self.assertEqual(first,second);kill.assert_not_called();start.assert_not_called()
        self.assertEqual(first['context_length'],400000);self.assertEqual(first['concurrency'],1)
        self.assertEqual(m.read_json(self.destination/'omlx.json'),self.config)
        evidence=m.read_json(self.destination/'evidence.json')
        self.assertEqual(evidence['stable_observations'],3)
        self.assertEqual(first['evidence_sha256'],m.omlx.fingerprint(evidence))
        for name in ['omlx.json','evidence.json','capture.lock']:
            self.assertEqual((self.destination/name).stat().st_mode&0o777,0o600)

    def test_mid_capture_identity_metadata_or_listener_change_never_installs(self):
        for change in ['instance','profile','listener','stopped','metadata']:
            with self.subTest(change=change):
                self.destination=self.root/change
                count=[0]
                def inspect(_):
                    count[0]+=1;value=copy.deepcopy(self.inspection)
                    if count[0]>1 and change!='metadata':value[change]=False if change=='listener' else True if change=='stopped' else 'd'*32
                    return value
                reads=[0]
                def metadata(_):
                    reads[0]+=1;return {**self.metadata,**({'context_length':8192} if reads[0]>1 and change=='metadata' else {})}
                with self.assertRaisesRegex(ValueError,'no_longer_current'):self.run_capture(inspect=inspect,metadata=metadata)
                self.assertFalse((self.destination/'omlx.json').exists())

    def test_resumed_capture_cannot_replace_pinned_configuration_or_process(self):
        self.run_capture();original=(self.destination/'evidence.json').read_bytes()
        self.config['command_sha256']='d'*64
        with self.assertRaisesRegex(ValueError,'existing_enrollment_changed'):self.run_capture()
        self.config['command_sha256']='a'*64;self.inspection['instance']='d'*32
        with self.assertRaisesRegex(ValueError,'existing_capture_changed'):self.run_capture()
        self.assertEqual((self.destination/'evidence.json').read_bytes(),original)

    def test_readonly_partial_capture_can_finish_after_metadata_write(self):
        self.run_capture();(self.destination/'omlx.json').unlink()
        old=(self.destination/'evidence.json').read_bytes();self.run_capture()
        self.assertEqual((self.destination/'evidence.json').read_bytes(),old)
        self.assertTrue((self.destination/'omlx.json').exists())

    def test_unsafe_paths_and_changed_private_inputs_refuse_before_capture(self):
        for change in [{'route':{'id':'local','url':'http://127.0.0.1:9999/v1'}},
                       {'launcher':'relative'}, {'profile_files':['/a','/a']}, {'concurrency':True}, {'extra':'arbitrary'}]:
            with self.subTest(change=change),self.assertRaises(ValueError):m.validate_expected({**self.expected,**change})
        for url in ['https://127.0.0.1:8013/v1','http://example.invalid:8013/v1','http://user@127.0.0.1:8013/v1','http://127.0.0.1:8013/v1?x=1']:
            changed=copy.deepcopy(self.expected);changed['route']['url']=changed['target']['url']=url
            with self.assertRaises(ValueError):m.validate_expected(changed)

    def test_existing_public_or_symlinked_output_is_not_trusted(self):
        self.run_capture();wrapper=self.destination/'omlx.json';wrapper.chmod(0o644)
        with self.assertRaisesRegex(ValueError,'file_unverified'):self.run_capture()
        wrapper.unlink();wrapper.symlink_to(self.destination/'evidence.json')
        with self.assertRaisesRegex(ValueError,'file_unverified'):self.run_capture()


if __name__=='__main__':unittest.main()

File: ds4-gateway/recovery-omlx-enrollment.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import {randomUUID} from 'node:crypto';
import {createOmlxEnrollment,restoreOmlxEnrollments} from './recovery-omlx-enrollment.mjs';
import {Recovery} from './recovery.mjs';
import {createRecoveryTools} from './genie-recovery.mjs';
import http from 'node:http';
import {createGateway} from './gateway.mjs';
import {workerControl} from './worker-client.mjs';

function fixture(t,{cleanup=true}={}){
  const directory=fs.mkdtempSync(path.join(os.tmpdir(),'omlx-enroll-'));if(cleanup)t.after(()=>fs.rmSync(directory,{recursive:true,force:true}));
  const worker={id:'custom-local',url:'http://127.0.0.1:8013/v1',backend:'openai',api_key_file:'/fixture/inference-key',model_aliases:{PoolModel:'fixture'},contextLength:400000,max_concurrent_requests:1,healthy:true,drained:false,active:null,queue:[]};
  const config={model:'PoolModel',state_file:path.join(directory,'state.json'),control_socket:path.join(directory,'control.sock'),genie_chat:{python:'/usr/bin/python3',inspection:{workers:{'custom-local':{kind:'omlx-local',root:'/fixture/omlx',url:worker.url,api_key_file:'/fixture/inspection-key'}}}},
    omlx_recovery_setup:{workers:{'custom-local':{exclusive:true,launcher:'/fixture/omlx/start.sh',profile_files:['/fixture/guard.sh']}}},machine_groups:{'custom-local':['one-machine']}};
  const store={filename:config.state_file,data:{recovery:{version:1,automatic:true,operations:[]}},save(value){fs.writeFileSync(this.filename,JSON.stringify(value));this.data=structuredClone(value);}};store.save(store.data);
  const recovery=new Recovery({}, {store,nodes:[worker],model:'PoolModel',stopping:()=>false,reinstate(){assert.fail('routing changed');},fleetConfig:config});
  let enabled=true;const calls=[];
  const result={machine:'a'.repeat(64),profile:'b'.repeat(64),evidence_sha256:'c'.repeat(64),instance:'d'.repeat(32),config_sha256:'e'.repeat(64),context_length:400000,concurrency:1};
  const options={config,store,recovery,isEnabled:()=>enabled,materialize:async input=>{calls.push(input);return result;}};
  return {directory,config,store,recovery,worker,calls,result,options,enabled:v=>enabled=v,request:()=>({worker_id:worker.id,action_id:randomUUID()}),create:()=>createOmlxEnrollment(options)};
}
test('native local enrollment retains backups, exact inputs and legacy services across reconstruction',async t=>{
  const f=fixture(t),s=f.create(),input=f.request();assert.equal(s.request(input).state,'queued');await s.idle();
  const row=f.store.data.omlx_recovery_enrollments[input.action_id];assert.equal(row.state,'enrolled');
  assert.ok(fs.existsSync(row.backup));assert.ok(fs.existsSync(row.intent_backup));assert.equal(f.worker.drained,false);
  assert.equal(row.entry.verification,'glm53_omlx');assert.equal(row.entry.start_stopped,undefined);
  assert.equal(f.recovery.state.operations.length,0);assert.equal(s.request(input).state,'enrolled');assert.equal(f.calls.length,1);
  assert.deepEqual(f.calls[0].expected.target,f.config.genie_chat.inspection.workers[f.worker.id]);assert.equal(f.calls[0].expected.concurrency,1);
  const fresh={};restoreOmlxEnrollments(fresh,f.store.data.omlx_recovery_enrollments,[f.worker]);assert.deepEqual(fresh.recovery.workers,[row.entry]);
  assert.ok(!JSON.stringify(s.status()).includes('/fixture'));assert.equal(s.status().operations[0].entry,undefined);
});
test('explicit policy, exact inspection endpoint, ownership and healthy unpaused worker gate capture',async t=>{
  for(const change of [f=>f.enabled(false),f=>f.worker.drained=true,f=>f.worker.active={},f=>f.worker.url='http://127.0.0.1:9999/v1',
    f=>f.config.omlx_recovery_setup.workers[f.worker.id].launcher='relative',f=>f.recovery.configs.set(f.worker.id,{}),
    f=>f.config.machine_groups[f.worker.id]=['a','b'],f=>f.config.omlx_recovery_setup.workers[f.worker.id].start_stopped=true]){
    const f=fixture(t),s=f.create();change(f);assert.throws(()=>s.request(f.request()));assert.equal(f.calls.length,0);assert.equal(s.status().operations.length,0);
  }
});
test('late policy, credential, capacity or ownership change cannot install captured authority',async t=>{
  for(const change of [f=>f.enabled(false),f=>f.worker.api_key_file='/different/key',f=>f.worker.contextLength=8192,f=>f.worker.active={},f=>f.worker.drained=true]){
    const f=fixture(t);f.options.materialize=async()=>{change(f);return f.result;};const s=f.create();s.request(f.request());await s.idle();
    assert.equal(s.status().operations[0].state,'failed');assert.equal(f.recovery.config(f.worker.id),undefined);
  }
});
test('graceful replacement preserves queued identity and resumes only the same read-only capture',async t=>{
  const f=fixture(t);let finish;f.options.materialize=()=>new Promise(resolve=>finish=resolve);
  const before=f.create(),input=f.request();before.request(input);before.close();finish(f.result);await before.idle();
  assert.equal(before.status().operations[0].state,'queued');assert.equal(f.recovery.config(f.worker.id),undefined);
  f.options.materialize=async()=>f.result;const after=f.create();after.tick();await after.idle();
  assert.equal(after.status().operations[0].action_id,input.action_id);assert.equal(after.status().operations[0].state,'enrolled');
  assert.throws(()=>after.request(f.request()),/already_requested/);assert.throws(()=>after.request({...input,worker_id:'other'}),/conflict/);
});
test('native or metadata persistence failure preserves old configuration without leaking diagnostics',async t=>{
  const f=fixture(t);f.options.materialize=async()=>{throw Error('private native details');};const s=f.create();s.request(f.request());await s.idle();
  assert.equal(s.status().operations[0].error,'omlx_enrollment_unverified');assert.equal(f.recovery.config(f.worker.id),undefined);
  const g=fixture(t);g.options.materialize=async()=>{g.store.save=()=>{throw Error('disk full');};return g.result;};const next=g.create();next.request(g.request());await next.idle();
  assert.equal(next.status().operations[0].state,'queued');assert.equal(g.recovery.config(g.worker.id),undefined);
});
test('restore rejects collisions, changed adapter and invented stopped-start authority',async t=>{
  const f=fixture(t),s=f.create();s.request(f.request());await s.idle();const saved=f.store.data.omlx_recovery_enrollments;
  assert.throws(()=>restoreOmlxEnrollments({recovery:{workers:[Object.values(saved)[0].entry]}},saved,[f.worker]),/one registered/);
  for(const change of [{adapter:'docker'},{start_stopped:true},{verification:'qwen_omlx'}]){
    const bad=structuredClone(saved);Object.assign(Object.values(bad)[0].entry,change);assert.throws(()=>restoreOmlxEnrollments({},bad,[f.worker]),/journal_invalid/);
  }
});
test('Genie local enrollment accepts only the worker and its durable action ID under all capability gates',async()=>{
  let enabled=true,changes=true,inspection=true,testing=false;const calls=[];
  const q=createRecoveryTools({read:async()=>({version:1,recovery:{workers:[],operations:[]}}),enrollOmlx:async input=>{calls.push(input);return {...input,state:'queued'};},isEnabled:()=>enabled,isChangesEnabled:()=>changes,isInspectionEnabled:()=>inspection,isTesting:()=>testing});
  const input={action:'enroll-omlx',action_id:randomUUID(),worker_id:'custom'};
  assert.equal((await q.tool(input)).state,'queued');await assert.rejects(q.tool({...input,launcher:'/untrusted'}),/Specify/);
  for(const toggle of [()=>enabled=false,()=>{enabled=true;changes=false;},()=>{changes=true;inspection=false;},()=>{inspection=true;testing=true;}]){toggle();await assert.rejects(q.tool(input),/suspended/);}
  assert.equal(calls.length,1);
});
test('real core socket durably queues one local enrollment and refuses absent native files without actions',async t=>{
  const f=fixture(t,{cleanup:false}),server=http.createServer((_req,res)=>res.end(JSON.stringify({data:[{id:'fixture',max_model_len:400000}]})));let gateway;
  t.after(async()=>{try{await gateway?.close();}finally{server.closeAllConnections();await new Promise(resolve=>server.close(resolve));fs.rmSync(f.directory,{recursive:true,force:true});}});
  await new Promise(resolve=>server.listen(0,'127.0.0.1',resolve));const url=`http://127.0.0.1:${server.address().port}/v1`;
  const config={...f.config,host:'127.0.0.1',port:0,api_key:'none',context_length:400000,probe_timeout_ms:1000,health_interval_ms:60000,
    nodes:[{id:'custom-local',url,backend:'openai',model_aliases:{PoolModel:'fixture'},max_concurrent_requests:1}]};config.genie_chat.inspection.workers['custom-local'].url=url;
  fs.unlinkSync(config.state_file);gateway=createGateway(config);await gateway.start();const input=f.request();
  await assert.rejects(workerControl(config.control_socket,'/enroll-omlx-recovery',input),/policy_disabled/);
  await workerControl(config.control_socket,'/recovery-policy',{enabled:true});assert.equal((await workerControl(config.control_socket,'/enroll-omlx-recovery',input)).state,'queued');
  let row;const deadline=Date.now()+30000;
  do{row=(await workerControl(config.control_socket,'/workers')).recovery.omlx_enrollment.operations.find(r=>r.action_id===input.action_id);if(row?.state!=='queued')break;await new Promise(resolve=>setTimeout(resolve,50));}while(Date.now()<deadline);
  assert.equal(row.state,'failed');assert.equal(gateway.recovery.config('custom-local'),undefined);
  assert.equal((await workerControl(config.control_socket,'/enroll-omlx-recovery',input)).state,'failed');
  assert.equal((await workerControl(config.control_socket,'/workers')).recovery.omlx_enrollment.operations.length,1);
});
